tools/generate_db_schema.py

```diff
@@ -0,0 +1,227 @@
+    table_name_to_id_column_name = {}
+    for cls in primary_table_classes:
+        schema += f"table {cls.table_name}{{\n"
+        # If the class has a `parent_table_class`, then it is an expansion of a parent
+        # class and the id from the parent class is re-used
+        if cls.parent_table_class:
+            if isinstance(cls.parent_table_class, tuple):  # Multiple inheritance case
+                for parent_class in cls.parent_table_class:
+                    id_column_name = table_name_to_id_column_name[
+                        parent_class.table_name
+                    ]
+                    schema += (
+                        f"  {parent_class.table_name.rstrip('s')}_id {_id_type} "
+                        f"[pk, ref: - {parent_class.table_name}."
+                        f"{id_column_name}]\n"
+                    )
+            else:  # Single super class
+                id_column_name = f"{cls.parent_table_class.table_name.rstrip('s')}_id"
+                table_name_to_id_column_name[cls.table_name] = id_column_name
+                schema += (
+                    f"  {id_column_name} {_id_type} "
+                    f"[pk, ref: - {cls.parent_table_class.table_name}.id]\n"
+                )
+        else:
+            table_name_to_id_column_name[cls.table_name] = "id"
+
+        for column in cls.columns:
+            schema += f"  {column.name} {_type_translation[column.ctype]}"
+            if column.name == "id":
+                schema += " [pk, increment]\n"
+            else:
+                schema += "\n"
+        schema += "}\n\n"
+
+    return schema, table_name_to_id_column_name
+
+
+def _generate_DBML_for_linker_tables(
+    linker_tables, primary_table_classes, table_name_to_id_column_name
+):
+    """Return DBML for `linker_tables`, using also the set of `primary_table_classes` and
+    the table_name -> id_column_name mapping
+
+    """
+    schema = ""
+    # Generate schema for linker tables
+    table_name_to_primary_class = {cls.table_name: cls for cls in primary_table_classes}
+    for parent_class, owned_object_list in linker_tables:
+        # Broken, so skip for now
+        if "component" in owned_object_list.joining_table_name:
+            continue
+
+        # Owner foreign key
+        schema += f"table {owned_object_list.joining_table_name}{{\n"
+
+        def link(cls_):
+            """Return column DBML for link class"""
+            # Determin the name of the column the foreign key points to
+            foreign_key_column_name = table_name_to_id_column_name[cls_.table_name]
+            if foreign_key_column_name == "id":
+                column_name = cls_.table_name.rstrip("s") + "_id"
+            else:
+                column_name = foreign_key_column_name
+
+            # Generate link text on the form:
+            #   calibration_id INTEGER [pk, ref: > ms_calibrations.calibration_id]
+            link_text = (
+                f"  {column_name} {_id_type} [pk, ref: > "
```

Comment:
You've used "ref: > " here but "ref: - " above (line 153). 

Is that intentional to imply a different type of relationship?

Reply:
Yes, in the DBML that's the difference between the relationship. I think the - is one to one and the < and > is one-to-many where the direction of the arrow indicates which is which. More about that here: https://www.dbml.org/docs/#relationships-foreign-key-definitions